crates/bevy_ecs/src/system/query.rs

```diff
@@ -382,23 +390,18 @@ impl<D: QueryData, F: QueryFilter> core::fmt::Debug for Query<'_, '_, D, F> {
 impl<'w, 's, D: QueryData, F: QueryFilter> Query<'w, 's, D, F> {
     /// Creates a new query.
     ///
-    /// # Panics
-    ///
-    /// This will panic if the world used to create `state` is not `world`.
-    ///
     /// # Safety
     ///
-    /// This will create a query that could violate memory safety rules. Make sure that this is only
-    /// called in ways that ensure the queries have unique mutable access.
+    /// * This will create a query that could violate memory safety rules. Make sure that this is only
+    ///   called in ways that ensure the queries have unique mutable access.
+    /// * `world` must be the world used to create `state`.
     #[inline]
     pub(crate) unsafe fn new(
         world: UnsafeWorldCell<'w>,
         state: &'s QueryState<D, F>,
         last_run: Tick,
         this_run: Tick,
     ) -> Self {
-        state.validate_world(world.id());
```

Comment:
Why remove this check?

Reply:
Oh, right, that was subtle and I should have pointed it out!  

The reason is that I changed all of the `&self` and `&mut self` methods on `Query` to call `self.as_readonly().actual_method()` and `self.reborrow().actual_method()`, and `as_readonly()` and `reborrow()` are implemented in terms of `Query::new()`.  I wanted to make sure that change was zero-cost, but checking the assembly with `cargo-show-asm` showed that it was adding a call to `validate_world` to every existing method on `Query`!  Since we always had a world that was already validated, I changed it to be a safety requirement, and now the assembly appears to be unchanged.  

(Then I removed the test you asked about above because it's no longer true that it panics.)